File: Ai algorithm implementation/Random-Restart_Hill_Climbing.py
import random

N = 8

def heuristic(state):
    h = 0
    for i in range(N):
        for j in range(i + 1, N):
            if state[i] == state[j] or abs(state[i] - state[j]) == abs(i - j):
                h += 1
    return h

def hill_climb(state):
    while True:
        current_h = heuristic(state)
        best_state = state[:]
        best_h = current_h
        
        for col in range(N):
            original_row = state[col]
            for row in range(N):
                if row == original_row:
                    continue
                state[col] = row
                h = heuristic(state)
                if h < best_h:
                    best_h = h
                    best_state = state[:]
            state[col] = original_row
        
        if best_h >= current_h:
            break
        state = best_state[:]
    
    return state

def random_restart_hc(restarts=1000):
    best_state = None
    best_h = float('inf')
    
    for r in range(restarts):
        state = [random.randint(0, N - 1) for _ in range(N)]
        sol = hill_climb(state)
        h = heuristic(sol)
        if h < best_h:
            best_h = h
            best_state = sol
            if h == 0:
                break
    
    return best_state

random.seed()
solution = random_restart_hc()
print(f"Random-Restart Hill Climbing Solution (Heuristic={heuristic(solution)}):")
print(solution)

File: Ai algorithm implementation/Random-Restart_Hill_Climbing.cpp
#include <bits/stdc++.h>
using namespace std;

int N = 8;

int heuristic(vector<int> &state) {
    int h = 0;
    for(int i=0;i<N;i++)
        for(int j=i+1;j<N;j++)
            if(state[i]==state[j] || abs(state[i]-state[j])==abs(i-j)) h++;
    return h;
}

vector<int> hillClimb(vector<int> state) {
    while(true) {
        int currentH = heuristic(state);
        vector<int> bestState = state;
        int bestH = currentH;

        for(int col=0; col<N; col++) {
            int originalRow = state[col];
            for(int row=0; row<N; row++) {
                if(row==originalRow) continue;
                state[col] = row;
                int h = heuristic(state);
                if(h < bestH) {
                    bestH = h;
                    bestState = state;
                }
            }
            state[col] = originalRow;
        }

        if(bestH >= currentH) break;
        state = bestState;
    }
    return state;
}

vector<int> randomRestartHC(int restarts=1000) {
    vector<int> bestState;
    int bestH = INT_MAX;
    srand(time(0));
    for(int r=0;r<restarts;r++) {
        vector<int> state(N);
        for(int i=0;i<N;i++) state[i] = rand()%N;

        vector<int> sol = hillClimb(state);
        int h = heuristic(sol);
        if(h < bestH) {
            bestH = h;
            bestState = sol;
            if(h==0) break;
        }
    }
    return bestState;
}

int main() {
    vector<int> solution = randomRestartHC();
    cout << "Random-Restart Hill Climbing Solution (Heuristic=" << heuristic(solution) << "):\n";
    for(int i=0;i<N;i++) cout << solution[i] << " ";
    cout << "\n";
}
